indication can be a bit pattern (#1)
mjoras/trone: draft-thoji-scone-trone-protocol.md

# TRONE Indication {#trone-indication}

QUIC endpoints can signal potential support for TRONE before the completion of
the QUIC handshake by sending a QUIC long header packet as the first UDP
datagram with its last 16 bytes set to the TRONE Indication Pattern.

The TRONE Indication Pattern is "givemesomescone!" encoded in US-ASCII.

The TRONE indication provides an opportunistic signal to network elements that
the client might support TRONE. Network elements can use this as an early hint,
but must await confirmation of TRONE support by observing a full TRONE packet
after the handshake completes.

## Sending TRONE Indication

When indicating support for TRONE, a QUIC endpoint sets the last 16 bytes of the
first UDP datagram it sends to the TRONE Indication Pattern. This is an
opportunistic signal that is not guaranteed to be processed by the receiver, as
the receiver has not yet confirmed its willingness to send TRONE packets.

> In the case of QUIC v1, endpoints can simply append the TRONE Indication Pattern
> to the last QUIC long header packet of a UDP datagram. When such a UDP datagram

Reviewer: To the last QUIC packet or after it?

```suggestion
In the case of QUIC v1, endpoints can simply append the TRONE Indication Pattern (TIP)
after the last QUIC long header packet of a UDP datagram. When such a UDP datagram
```

Author: Thank you for the review comments!

is received by an endpoint that does not understand TRONE, the TRONE Indication
Pattern would be ignored while the long header packets will be processed
({{Section 12.2 of QUIC}}).

Other versions of QUIC might not support TRONE, or require different measures to

> set the last 16 bytes to the TRONE Indication Pattern.

Reviewer: This makes me more concerned about the viability of this approach than anything else.  This TIP only works if all packets are understood enough to ignore them AND know their length.  We have an existence proof of that not working in QUIC v1 (the short header packets) and there is no guarantee that other QUIC versions will retain a usable length.

steps:

1. Identify a UDP datagram containing a QUIC long header packet as defined in
   {{INVARIANTS}}, and that the size of Version-Specific Data is no less than
   16 bytes.
2. Examine the last 16 byte of the datagram to see if it has the TRONE
   Indication Bit Pattern.

The following pseudocode shows how a network element might detect a TRONE
indication:

~~~ pseudocode
is_long_header_quic = is_quic_long_header_packet(datagram)
if is_long_header_quic
  version_specific_data = get_quic_version_specific_data(datagram)
  if length(version_specific_data) >= 16
    potential_bit_pattern = datagram[length(datagram) - 16,length(datagram)]
    if potential_bit_pattern == TRONE_INDICATION_BIT_PATTERN
      note_potential_trone_support(flow_tuple)
~~~

This approach allows network elements to detect TRONE support without needing
to parse the version-specific QUIC packet structure or perform deep packet
inspection of the TLS handshake.

Network elements that observe TRONE indication packets MAY note the UDP 4-tuple
for potential future TRONE handling.

Network elements MUST NOT rely solely on the presence of a TRONE indication to

> confirm that a flow supports TRONE. A flow SHOULD only be confirmed as

Reviewer: I think that this is "is", not "SHOULD":

```suggestion
confirm that a flow supports TRONE. A flow is only confirmed as
```
